Add a phase to perform functional to inplace operator conversion

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57477 [JIT] Add a phase to perform functional to inplace operator conversion**

Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

desertfire committed:
[JIT] Add a phase to perform inplace<->functional conversion for activation operators

Summary: Currently the conversion only deals with activation operators. The legality check is somewhat strict for now.

Test Plan:
```
python test/test_jit.py -k test_functional_to_inplace_activation
python test/test_jit.py -k test_inplace_to_functional_activation
```

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add a phase to perform inplace<->functional conversion for activation operators"


Summary: Currently the conversion only deals with activation operators. The legality check is somewhat strict for now.

Test Plan:
```
python test/test_jit.py -k test_functional_to_inplace_activation
python test/test_jit.py -k test_inplace_to_functional_activation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

penguinwu commented on test/jit/test_convert_activation.py:
```diff
@@ -0,0 +1,82 @@
+            torch.nn.functional.hardtanh,
+            torch.nn.functional.hardswish,
+            torch.nn.functional.relu,
+            torch.nn.functional.relu6,
```
why is `sigmoid` and `tanh` included in the `TestFunctionalToInplaceActivation` but not `TestInplaceToFunctionalActivation`?

If we are supposed to test exactly the same set of activation functions for both tests, then perhaps to define a global `ActivationList` that is used by both tests would avoid such asymmtry.

desertfire replied:
torch.sigmoid or torch.tanh use a different syntax for its inplace version. I will rewrite this test.

penguinwu commented on torch/csrc/jit/passes/convert_activation.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    // If x is an input parameter or alias with an input parameter, skip the
+    // conversion
+    if (input->node()->kind() == prim::Param ||
```
what if the input is a weight tensor? Basically, in model execution, are there other types of tensors (other than param) whose live-range exceeds the current inference execution?

I think the criteria here is to make sure that the input is a last-use. There are tensors who are alive throughout the inference (e.g., graph inputs or weights).

penguinwu commented on torch/csrc/jit/passes/convert_activation.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    // Use y = relu(x) as an example:
+    // If y and x have different type, skip the conversion
+    if (input->type() != output->type()) {
```
What if the input and output sizes differ?

eellison replied:
So, this check is getting at the fact that we can't inplace a node if the output dtype is separate from the input dtype. Could you comment that ? Shape isn't necessary for activation since the size will always remain unchanged. 

If we're going to use `input->type() != output->type()` to verify that the output dtype equals the input dtype, we also need to check that the dtype is actually specified on the types. It is not guaranteed to be. 

I would maintain a list of activations you know do not type promote (relu for instance), and then for the ones not in that list that do type promote (sigmoid), check that the dtype is specified on both input and output and they are equal.

penguinwu commented on torch/csrc/jit/passes/convert_activation.cpp:
```diff
@@ -0,0 +1,100 @@
+    Value* input = node->inputs().at(0);
+    Value* output = node->output();
+
+    // Use y = relu(x) as an example:
```
The key part of the algorithm is to decide which activation functions can be rewritten to the in-place form. Could you carve out that part of the logic into a separate helper function and describe the algorithm in the comments? I feel that the current criteria is safe but not optimal and may be improved later and labeling this part of the algorithm separately may make it easier for future update

penguinwu commented on torch/csrc/jit/passes/convert_activation.h:
```diff
@@ -0,0 +1,40 @@
+namespace jit {
+
+class FunctionalActivationRewriter {
+ public:
```
Please add comments to describe what the passes do

penguinwu reviewed (changes requested):
I have a couple of questions regarding the algorithm that checks whether a functional op can be replaced by functional op. Please check those.

eellison commented on torch/csrc/jit/passes/convert_activation.cpp:
```diff
@@ -0,0 +1,100 @@
+    changed = true;
+    Node* inplace_node = node->replaceWithNewSymbol(
+        Symbol::fromQualString(node->schema().name() + "_"));
+    inplace_node->output()->replaceAllUsesWith(input);
```
Even through using the input or output are functionally equivalent, it is more idiomatic to use the output.

desertfire replied:
> Even through using the input or output are functionally equivalent, it is more idiomatic to use the output.

Are you suggesting `inplace_node->input()->replaceAllUsesWith(output)`? I personally found that a little counter-intuitive.

eellison commented on torch/csrc/jit/passes/convert_activation.cpp:
```diff
@@ -0,0 +1,100 @@
+    Node* inplace_node = node->replaceWithNewSymbol(
+        Symbol::fromQualString(node->schema().name() + "_"));
+    inplace_node->output()->replaceAllUsesWith(input);
+    getOrCreateAliasDb()->replaceWithNewValue(output, input);
```
Can we do `getOrCreateAliasDb()->replaceWithNewValue(node->output, inplace_node->output());` ?

eellison reviewed (commented):
Looks good, I think we need to update the logic around when it is safe to reinplace nodes around type promotion.

desertfire committed:
Update on "[JIT] Add a phase to perform inplace<->functional conversion for activation operators"


Summary: Currently the conversion only deals with activation operators. The legality check is somewhat strict for now.

Test Plan:
```
python test/test_jit.py -k test_functional_to_inplace_activation
python test/test_jit.py -k test_inplace_to_functional_activation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

eellison commented on torch/csrc/jit/passes/restore_mutation.cpp:
```diff
@@ -0,0 +1,136 @@
+  //  return false;
+  //}
+
+  // In general, we don't need to check shape as activation ops are
```
We're not checking that there's an equivalent inplace schema for the node when making this transformation, which we know for activations but we don't know if we're making this a public exposed api

desertfire replied:
I will add a schema checking.

eellison reviewed (commented):
Looking good! One comment is that i'm not really sure if opening this up as a public API makes sense... I don't know what the intended use case is and the extra abstractions are making the pass more complicated. If we are opening this up as a public API we need to make sure that the conditions are correct. 

The other comment is that the operators we can reinplace are the operators that dont type promote or dont accept integer inputs. if you look at [this script](https://gist.github.com/eellison/aa67f0f13d3205538e77acfb22dd397d) I think it's a greater set than you have. We might consider checking in some version of this script as a test so that if integer inputs are added or type promotion is added to these operators the test will fail. 

`selu
hardsigmoid
hardtanh
hardswish
rrelu_
elu
relu
leaky_relu
celu
relu6
silu`

desertfire commented on torch/csrc/jit/passes/restore_mutation.h:
```diff
@@ -0,0 +1,58 @@
+// so that we won't lose the performance benefit coming from memory reduction.
+
+// Replaces in-place aten activation ops with their functional equivalents
+TORCH_API bool InplaceToFunctionalActivation(
```
Initially I was thinking to put it in remove_mutation.cpp, but that requires we expose the activation op set as global. I don't see many other passes are exposing this kind of global data, but if that is acceptable, I can change the code.

desertfire replied:
We probably need a centralized location to store the activation op sets, or is it possible to add attributes (e.g. is_elementwise, is_type_promotable, ..) for our builtin ops?

desertfire commented on test/jit/test_convert_activation.py:
```diff
@@ -0,0 +1,137 @@
+
+class TestInplaceToFunctionalActivation(JitTestCase):
+    def test_inplace_to_functional_activation(self):
+        for activation in [
```
Because the activation syntax are different for the two groups, `_` vs `inplace=True`

desertfire commented on torch/csrc/jit/passes/restore_mutation.h:
```diff
@@ -0,0 +1,58 @@
+namespace torch {
+namespace jit {
+
+struct FunctionalToInplaceRewriterFlags {
```
I was thinking to use this to later support more builtin ops beyond activation ops. For example, if later on we want to support add ->add_ , we want to check the shape after the shape analysis is more mature, but we still won't need to check shape for element-wise ops. I guess a cleaner way to implement this is to associate this kind of information to each group of ops.

eellison replied:
Implementing it for add would add a dependency on dtype/shape analysis, I don't think this API would be part of it.

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

eellison commented on test/jit/test_convert_activation.py:
```diff
@@ -0,0 +1,175 @@
+            try:
+                out = activation(inp)
+                if dtype != out.dtype:
+                    exit()
```
nit: i'm unfamiliar with exit, never seen it before...

eellison commented on test/jit/test_convert_activation.py:
```diff
@@ -0,0 +1,175 @@
+            torch.float32,
+            torch.float64,
+        ]
+        # restore_mutation.h contains a mapping from activation operators
```
nice !

eellison reviewed (approved):
looks great 😍 😍 😍 great job!!

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).

desertfire committed:
Update on "[JIT] Add a phase to perform functional to inplace operator conversion"


Summary: This diff contains two parts:
1. Inplace to functional conversion by calling existing RemoveTensorMutation.
2. A new pass to convert from functional to inplace. Currently the conversion only
deals with activation operators. The legality check is restricted until we have
dtype and shape analysis.

Test Plan:
```
python test/test_jit.py -k TestFunctionalToInplaceActivation
python test/test_jit.py -k TestInplaceToFunctionalActivation
```

Differential Revision: [D28155153](https://our.internmc.facebook.com/intern/diff/D28155153)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28155153).
